File: Recursion/Combination_sum.cpp
#include <bits/stdc++.h>
using namespace std;

void find_combo(int ind, int arr[], int n, int target, vector<vector<int>> &ans, vector<int> &ds) {
    if (ind == n) {
        if (target == 0) {
            ans.push_back(ds);
        }
        return;
    }

    // Pick the element if it's <= target
    if (arr[ind] <= target) {
        ds.push_back(arr[ind]);
        find_combo(ind, arr, n, target - arr[ind], ans, ds);  // same index because repetition allowed
        ds.pop_back();
    }

    // Skip the element
    find_combo(ind + 1, arr, n, target, ans, ds);
}

int main() {
    int arr[] = {5, 3, 4, 1};
    int n = sizeof(arr) / sizeof(arr[0]);
    int target = 4;

    vector<int> ds;
    vector<vector<int>> ans;

    find_combo(0, arr, n, target, ans, ds);

    // Print all combinations
    for (auto &v : ans) {
        for (int num : v)
            cout << num << " ";
        cout << endl;
    }

    return 0;
}

File: java/Recursion/CombinationSum.java
package Recursion;
import java.util.*;

public class CombinationSum{

    /*
    ====================================================
    INTUITION:
    ----------------------------------------------------
    - At every index, we have two choices:
        1) PICK the element (if it does not exceed target)
           → stay on the same index (repetition allowed)
        2) NOT PICK the element
           → move to next index
    - When we reach index == n:
        - If target == 0, we found a valid combination
    ====================================================
    */

    static void findCombo(int ind, int[] arr, int n, int target,
                          List<List<Integer>> ans, List<Integer> ds) {

        // Base case
        if (ind == n) {
            if (target == 0) {
                ans.add(new ArrayList<>(ds)); // store copy
            }
            return;
        }

        // PICK the element (repetition allowed)
        if (arr[ind] <= target) {
            ds.add(arr[ind]);
            findCombo(ind, arr, n, target - arr[ind], ans, ds);
            ds.remove(ds.size() - 1); // backtrack
        }

        // NOT PICK the element
        findCombo(ind + 1, arr, n, target, ans, ds);
    }

    public static void main(String[] args) {

        int[] arr = {5, 3, 4, 1};
        int n = arr.length;
        int target = 4;

        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> ds = new ArrayList<>();

        findCombo(0, arr, n, target, ans, ds);

        // Print all combinations
        for (List<Integer> list : ans) {
            for (int num : list) {
                System.out.print(num + " ");
            }
            System.out.println();
        }
    }
}
